scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala

```diff
@@ -976,15 +988,28 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
         case tpe: Type.Singleton => tpe
         case Type.Apply(underlying, args) => atPos(tpe, tpe)(Pat.Type.Apply(loop(underlying), args.map(convert)))
         case Type.ApplyInfix(lhs, op, rhs) => atPos(tpe, tpe)(Pat.Type.ApplyInfix(loop(lhs), op, loop(rhs)))
-        case Type.Function(params, res) => atPos(tpe, tpe)(Pat.Type.Function(params.map(p => loop(p.require[Type])), loop(res)))
+        case Type.Function(params, res) => atPos(tpe, tpe)(Pat.Type.Function(params.map {
+          case q @ Type.Arg.Quasi(1, Type.Arg.Quasi(0, _)) => q.become[Pat.Type.Quasi]
+          case p => loop(p.require[Type])
+        }, loop(res)))
         case Type.Tuple(elements) => atPos(tpe, tpe)(Pat.Type.Tuple(elements.map(loop)))
         case Type.Compound(tpes, refinement) => atPos(tpe, tpe)(Pat.Type.Compound(tpes.map(loop), refinement))
         case Type.Existential(underlying, quants) => atPos(tpe, tpe)(Pat.Type.Existential(loop(underlying), quants))
         case Type.Annotate(underlying, annots) => atPos(tpe, tpe)(Pat.Type.Annotate(loop(underlying), annots))
+        case Type.Placeholder(Type.Bounds(None, None)) => convert(tpe)
         case Type.Placeholder(bounds) => atPos(tpe, tpe)(Pat.Type.Placeholder(bounds))
         case tpe: Lit => tpe
       }
-      loop(compoundType())
+      val t: Type = {
+        val t = if (token.is[`(`]) tupleInfixType() else compoundType()
+        def mkOp(t1: Type) = atPos(t, t1)(Type.ApplyInfix(t, typeName(), t1))
+        token match {
+          case _: `forSome` => next(); atPos(t, t)(Type.Existential(t, existentialStats()))
+          case _: Unquote | _: Ident if !isRawBar => infixTypeRest(mkOp(compoundType()), InfixMode.LeftOp)
+          case _ => t
+        }
+      }
+      loop(t)
```

Comment:
This method didn't work correctly with `(` as token, `forSome`, and probably something else.